File: backend/app/schemas/battle.py
"""
배틀 WebSocket 메시지 스키마

클라이언트 → 서버 (C→S):
  join_queue        : 매칭 큐 진입
  leave_queue       : 큐 이탈
  heartbeat         : 5초마다 생존 신호
  typing_progress   : 타이핑 진행도 업데이트 (배틀 중)
  finish            : 텍스트 완료 선언

서버 → 클라이언트 (S→C):
  queued            : 큐 진입 완료
  queue_position    : 큐 내 위치 / 예상 대기시간
  matched           : 매칭 성사 (상대 정보 포함)
  countdown         : 카운트다운 숫자
  game_start        : 게임 시작 (공유 텍스트 전달)
  opponent_progress : 상대방 진행도
  game_end          : 게임 종료 결과
  opponent_left     : 상대방 연결 끊김
  error             : 에러 메시지
"""
from __future__ import annotations
from pydantic import BaseModel


# ── 공용 ──────────────────────────────────────────────────────────────

class OpponentInfo(BaseModel):
    user_id:  str
    username: str
    mmr:      float
    level:    int


class GameResult(BaseModel):
    room_id:       str
    winner_id:     str | None      # None = 무승부
    your_cpm:      int
    opp_cpm:       int
    your_accuracy: int
    opp_accuracy:  int
    mmr_delta:     int             # 내 MMR 변화량 (+/-)
    new_rank_score: int


# ── C→S 메시지 ────────────────────────────────────────────────────────

class JoinQueueMsg(BaseModel):
    type: str = "join_queue"
    mode: str = "quick"           # "quick" | "custom"
    room_code: str | None = None  # custom 전용


class HeartbeatMsg(BaseModel):
    type: str = "heartbeat"


class TypingProgressMsg(BaseModel):
    type:          str = "typing_progress"
    room_id:       str
    correct_chars: int
    cpm:           int
    accuracy:      int


class FinishMsg(BaseModel):
    type:     str = "finish"
    room_id:  str
    cpm:      int
    accuracy: int


class LeaveQueueMsg(BaseModel):
    type: str = "leave_queue"

File: src/api/battle.ts
/**
 * 실시간 배틀 WebSocket 메시지 타입
 *
 * 서버 ↔ 클라이언트 프로토콜을 TypeScript로 선언한다.
 * 실제 WebSocket 연결 로직은 useBattleSocket 훅에서 처리한다.
 */

// ── 서버 → 클라이언트 ────────────────────────────────────────────────

export interface S_Queued {
  type: 'queued';
  mmr: number;
}

export interface S_Dequeued {
  type: 'dequeued';
}

export interface S_Matched {
  type: 'matched';
  room_id: string;
  opponent: {
    user_id:  string;
    username: string;
    level:    number;
  };
}

export interface S_Countdown {
  type: 'countdown';
  seconds: number;
}

export interface S_GameStart {
  type:    'game_start';
  room_id: string;
  text:    string;
}

export interface S_OpponentProgress {
  type:          'opponent_progress';
  correct_chars: number;
  cpm:           number;
  accuracy:      number;
}

export interface S_OpponentFinished {
  type:     'opponent_finished';
  cpm:      number;
  accuracy: number;
}

export interface S_GameEnd {
  type:           'game_end';
  room_id:        string;
  winner_id:      string | null;
  your_cpm:       number;
  opp_cpm:        number;
  your_accuracy:  number;
  opp_accuracy:   number;
  mmr_delta:      number;   // 내 MMR 변화량 (+/-)
}

export interface S_OpponentLeft {
  type: 'opponent_left';
}

export interface S_Error {
  type:    'error';
  message: string;
}

export type ServerMessage =
  | S_Queued
  | S_Dequeued
  | S_Matched
  | S_Countdown
  | S_GameStart
  | S_OpponentProgress
  | S_OpponentFinished
  | S_GameEnd
  | S_OpponentLeft
  | S_Error;

// ── 클라이언트 → 서버 ────────────────────────────────────────────────

export interface C_JoinQueue {
  type: 'join_queue';
  mode: 'quick' | 'custom';
  room_code?: string;
}

export interface C_LeaveQueue {
  type: 'leave_queue';
}

export interface C_Heartbeat {
  type: 'heartbeat';
}

export interface C_TypingProgress {
  type:          'typing_progress';
  room_id:       string;
  correct_chars: number;
  cpm:           number;
  accuracy:      number;
}

export interface C_Finish {
  type:     'finish';
  room_id:  string;
  cpm:      number;
  accuracy: number;
}

export type ClientMessage =
  | C_JoinQueue
  | C_LeaveQueue
  | C_Heartbeat
  | C_TypingProgress
  | C_Finish;
